alicloud/provider.go

```diff
@@ -305,6 +307,7 @@ func Provider() terraform.ResourceProvider {
 			"alicloud_key_pair":                      resourceAlicloudKeyPair(),
 			"alicloud_key_pair_attachment":           resourceAlicloudKeyPairAttachment(),
 			"alicloud_kms_key":                       resourceAlicloudKmsKey(),
+			"alicloud_kms_ciphertext":                resourceAlicloudKmsCiphertext(),
```

Comment:
I think there is no need to add a resource to get ciphertext

Reply:
The idea behind the ciphertext resource is that it stays stable across applies.
A data source re-encrypts the data on each apply, because it runs the Read method on each apply. A resource will only encrypt data when the parameters change.

So if you want to use the data source ciphertext in another resource (e.g. inside the user-data for a VM), terraform would replace that resource on every apply because the ciphertexts changes. If you would use the ciphertext resource in such a case, terraform would only replace the resource if a re-encryption is necessary.

Reply:
I think, for data source, if keyId and plaintext are not changed, the ciphertext will not be changed. Right?

Reply:
AFAIK the `Read` method of a data source is executed on every apply. So the plaintext will be re-encrypted on every apply. The encryption mechanism used by KMS is randomized (for increased security), so each generated ciphertext will be different, even if the key and plaintext stay the same.

So in effect terraform will detect a changed ciphertext on each apply.